Clean tries to drop a function from an extension and fails

_Original author: Martin.A...@gmail.com (May 11, 2012 10:11:31)_

We are using PostgreSQL 9.1 and are using the uuid-ossp extension. Clean tries to drop one of its functions. The extension has been installed manually as it required adding binaries to the DBMS.

Any way to work around this?

FlywayException: Error executing statement at line 64: DROP FUNCTION IF EXISTS &quot;public&quot;.&quot;uuid_nil&quot;() CASCADE
Occured in com.googlecode.flyway.core.migration.sql.SqlStatement in method execute, line number 78
Caused by org.postgresql.util.PSQLException: ERROR: cannot drop function uuid_nil() because extension uuid-ossp requires it
  Hint: You can drop extension uuid-ossp instead.
Occured in org.postgresql.core.v3.QueryExecutorImpl in method receiveErrorResponse, line number 2102

_Original issue: http://code.google.com/p/flyway/issues/detail?id=257_

This occurs with postgis as well, trying to clean yields:

```
Caused by: org.postgresql.util.PSQLException: ERROR: cannot drop view geography_columns because extension postgis requires it
```

Basically, extensions can create objects which are "special"

It is also not abnormal to have a DBA user which does NOT have permissions to load/unload extensions, but can execute DDL, used for migrations.

--- a/flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/postgresql/PostgreSQLSchema.java
+++ b/flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/postgresql/PostgreSQLSchema.java
@@ -15,13 +15,10 @@
  */
 package org.flywaydb.core.internal.dbsupport.postgresql;
 
-import org.flywaydb.core.api.FlywayException;
 import org.flywaydb.core.internal.dbsupport.JdbcTemplate;
 import org.flywaydb.core.internal.dbsupport.Schema;
 import org.flywaydb.core.internal.dbsupport.Table;
 import org.flywaydb.core.internal.dbsupport.Type;
-import org.flywaydb.core.internal.util.logging.Log;
-import org.flywaydb.core.internal.util.logging.LogFactory;
 
 import java.sql.SQLException;
 import java.util.ArrayList;
@@ -32,15 +29,13 @@ import java.util.Map;
  * PostgreSQL implementation of Schema.
  */
 public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
-    
-    private static final Log LOG = LogFactory.getLog(PostgreSQLSchema.class);
-    
+
     /**
      * Creates a new PostgreSQL schema.
      *
      * @param jdbcTemplate The Jdbc Template for communicating with the DB.
-     * @param dbSupport    The database-specific support.
-     * @param name         The name of the schema.
+     * @param dbSupport The database-specific support.
+     * @param name The name of the schema.
      */
     public PostgreSQLSchema(JdbcTemplate jdbcTemplate, PostgreSQLDbSupport dbSupport, String name) {
         super(jdbcTemplate, dbSupport, name);
@@ -117,13 +112,6 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
             jdbcTemplate.execute(statement);
         }
 
-        for (Type type : allTypes()) {
-            try {
-                type.drop();
-            } catch (FlywayException e) {
-                LOG.warn("Unable to drop type " + type + ": " + e.getClass() + "-" + e.getMessage());
-            }
-        }
     }
 
     /**
@@ -155,8 +143,8 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
     private List<String> generateDropStatementsForBaseTypes(boolean recreate) throws SQLException {
         List<String> typeNames =
                 jdbcTemplate.queryForStringList(
-                        "select typname from pg_catalog.pg_type where typcategory in ('P', 'U') and typnamespace in (select oid from pg_catalog.pg_namespace where nspname = ?)",
-                        name);
+                "select typname from pg_catalog.pg_type where typcategory in ('P', 'U') and typnamespace in (select oid from pg_catalog.pg_namespace where nspname = ?)",
+                name);
 
         List<String> statements = new ArrayList<String>();
         for (String typeName : typeNames) {
@@ -203,12 +191,12 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
     private List<String> generateDropStatementsForRoutines() throws SQLException {
         List<Map<String, String>> rows =
                 jdbcTemplate.queryForList(
-                        //Search for all functions
+                // Search for all functions
                         "SELECT proname, oidvectortypes(proargtypes) AS args "
                                 + "FROM pg_proc INNER JOIN pg_namespace ns ON (pg_proc.pronamespace = ns.oid) "
-                                //that don't depend on an extension
-                                + "LEFT JOIN pg_depend dep ON dep.objid = pg_proc.oid AND dep.deptype = 'e' "
-                                + "WHERE pg_proc.proisagg = false AND ns.nspname = ? AND dep.objid IS NULL",
+                // that don't depend on an extension
+                        + "LEFT JOIN pg_depend dep ON dep.objid = pg_proc.oid AND dep.deptype = 'e' "
+                        + "WHERE pg_proc.proisagg = false AND ns.nspname = ? AND dep.objid IS NULL",
                         name
                 );
 
@@ -286,9 +274,9 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
     private List<String> generateDropStatementsForViews() throws SQLException {
         List<String> viewNames =
                 jdbcTemplate.queryForStringList(
-                        //Search for all views
-                        "SELECT table_name FROM information_schema.views" +
-                        //that don't depend on an extension
+                // Search for all views
+                "SELECT table_name FROM information_schema.views" +
+                        // that don't depend on an extension
                         " LEFT JOIN pg_depend dep ON dep.objid = (table_schema || '.' || table_name)::regclass::oid AND dep.deptype = 'e'" +
                         " WHERE table_schema=? AND dep.objid IS NULL",
                 name);
@@ -304,18 +292,18 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
     protected Table[] doAllTables() throws SQLException {
         List<String> tableNames =
                 jdbcTemplate.queryForStringList(
-                        //Search for all the table names
-                        "SELECT t.table_name FROM information_schema.tables t" +
-                                //in this schema
-                                " WHERE table_schema=?" +
-                                //that are real tables (as opposed to views)
-                                " AND table_type='BASE TABLE'" +
-                                //and are not child tables (= do not inherit from another table).
+                // Search for all the table names
+                "SELECT t.table_name FROM information_schema.tables t" +
+                        // in this schema
+                        " WHERE table_schema=?" +
+                        // that are real tables (as opposed to views)
+                        " AND table_type='BASE TABLE'" +
+                        // and are not child tables (= do not inherit from another table).
                                 " AND NOT (SELECT EXISTS (SELECT inhrelid FROM pg_catalog.pg_inherits" +
                                 " WHERE inhrelid = (quote_ident(t.table_schema)||'.'||quote_ident(t.table_name))::regclass::oid))",
                         name
                 );
-        //Views and child tables are excluded as they are dropped with the parent table when using cascade.
+        // Views and child tables are excluded as they are dropped with the parent table when using cascade.
 
         Table[] tables = new Table[tableNames.size()];
         for (int i = 0; i < tableNames.size(); i++) {

--- a/flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/postgresql/PostgreSQLSchema.java
+++ b/flyway-core/src/main/java/org/flywaydb/core/internal/dbsupport/postgresql/PostgreSQLSchema.java
@@ -15,10 +15,13 @@
  */
 package org.flywaydb.core.internal.dbsupport.postgresql;
 
+import org.flywaydb.core.api.FlywayException;
 import org.flywaydb.core.internal.dbsupport.JdbcTemplate;
 import org.flywaydb.core.internal.dbsupport.Schema;
 import org.flywaydb.core.internal.dbsupport.Table;
 import org.flywaydb.core.internal.dbsupport.Type;
+import org.flywaydb.core.internal.util.logging.Log;
+import org.flywaydb.core.internal.util.logging.LogFactory;
 
 import java.sql.SQLException;
 import java.util.ArrayList;
@@ -29,6 +32,9 @@ import java.util.Map;
  * PostgreSQL implementation of Schema.
  */
 public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
+    
+    private static final Log LOG = LogFactory.getLog(PostgreSQLSchema.class);
+    
     /**
      * Creates a new PostgreSQL schema.
      *
@@ -112,7 +118,11 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
         }
 
         for (Type type : allTypes()) {
-            type.drop();
+            try {
+                type.drop();
+            } catch (FlywayException e) {
+                LOG.warn("Unable to drop type " + type + ": " + e.getClass() + "-" + e.getMessage());
+            }
         }
     }
 
@@ -193,9 +203,12 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
     private List<String> generateDropStatementsForRoutines() throws SQLException {
         List<Map<String, String>> rows =
                 jdbcTemplate.queryForList(
+                        //Search for all functions
                         "SELECT proname, oidvectortypes(proargtypes) AS args "
                                 + "FROM pg_proc INNER JOIN pg_namespace ns ON (pg_proc.pronamespace = ns.oid) "
-                                + "WHERE pg_proc.proisagg = false AND ns.nspname = ?",
+                                //that don't depend on an extension
+                                + "LEFT JOIN pg_depend dep ON dep.objid = pg_proc.oid AND dep.deptype = 'e' "
+                                + "WHERE pg_proc.proisagg = false AND ns.nspname = ? AND dep.objid IS NULL",
                         name
                 );
 
@@ -273,8 +286,12 @@ public class PostgreSQLSchema extends Schema<PostgreSQLDbSupport> {
     private List<String> generateDropStatementsForViews() throws SQLException {
         List<String> viewNames =
                 jdbcTemplate.queryForStringList(
-                        "SELECT table_name FROM information_schema.views WHERE table_schema=?", name);
-
+                        //Search for all views
+                        "SELECT table_name FROM information_schema.views" +
+                        //that don't depend on an extension
+                        " LEFT JOIN pg_depend dep ON dep.objid = (table_schema || '.' || table_name)::regclass::oid AND dep.deptype = 'e'" +
+                        " WHERE table_schema=? AND dep.objid IS NULL",
+                name);
         List<String> statements = new ArrayList<String>();
         for (String domainName : viewNames) {
             statements.add("DROP VIEW IF EXISTS " + dbSupport.quote(name, domainName) + " CASCADE");

